### PR TITLE
fix the namespaceScoped of cachers

### LINKED ISSUE #19939: Watch doesn't work for apis/extensions with namespace
- This works http://127.0.0.1:8080/apis/extensions/v1beta1/watch/deployments?resourceVersion=66 (used by controller manager) 
- This doesn't work (gives me no response) http://127.0.0.1:8080/apis/extensions/v1beta1/watch/namespaces/default/deployments?resourceVersion=66 (used by kubectl)

api/v1 works fine, but I can reproduce this for other apis/extensions resources (deployments, hpa, ...). 

cc @caesarxuchao @nikhiljindal @lavalamp

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #19939
It's strange that so many of them are misconfigured and haven't caused any problem so far.

@janetkuo

### CODE REVIEW

**Comment by lavalamp** [comment]
It *has* caused a problem-- someone else has this same exact change out for review, but I can't find it now.

In that review, I asked for it to use e.g. "pods.NamespaceScoped()" instead of "true", so that future copy-paste inheritence will be more likely to change it. But I can't find the PR now...

**Comment by caesarxuchao** [comment]
@lavalamp, I've addressed your comment. PTAL. Thanks.

**Comment by bprashanth** [comment]
Thanks for doing this, was on call so didn't get a chance to address the review comments (https://github.com/kubernetes/kubernetes/pull/19663) but you can go ahead and get yours in.

**Comment by liggitt** [comment]
any chance you could include a test that would have caught the error?

**Comment by smarterclayton** [comment]
If you can add the test I'll add LGTM

**Comment by lavalamp** [comment]
It's tricky to test this, since to be useful the test has to check resources that will be added in the future and that it therefore doesn't know about-- I suggested adding a check where REST objects are used that their storage & strategy scopes agree.

**Comment by caesarxuchao** [comment]
> I suggested adding a check where REST objects are used that their storage & strategy scopes agree.

@lavalamp, by actually doing this I found it perhaps is a bad idea. I pushed a second commit for the ease of discussion. See my comments in that commit for details.

**Comment by smarterclayton** [comment]
Instead of passing in the boolean to storage decorator, why not pass in the create strategy itself?  Then someone has to provide a valid strategy, and you can ensure you test that NamespaceScoped is true.  That would satisfy the test aspect for me.

**Comment by smarterclayton** [comment]
Or if not the create strategy, the `NamespaceScoped() bool` method as a sliced interface

**Comment by lavalamp** [comment]
> Or if not the create strategy, the NamespaceScoped() bool method as a sliced interface

Not sure what you mean by "sliced" but this sounds good to me.

**Comment by smarterclayton** [comment]
Sliced = a partial interface like:

```
type NamespaceScopeCheck interface {
  NamespaceScoped() bool
}
```

that create strategy satisfies

**Comment by caesarxuchao** [comment]
@smarterclayton yeah, I figured that out somehow. Could you take another look? Thanks.

**Comment by smarterclayton** [comment]
I feel more comfortable with this (you would have to provide a fake method, knowing that you have the strategy handy).  No further comments.

File: examples/apiserver/rest/reststorage.go [hunk 1]
```diff
@@ -63,3 +65,10 @@ func NewREST(s storage.Interface, storageDecorator generic.StorageDecorator) *RE
 	}
 	return &REST{store}
 }
+
+type NamespaceScopedStrategy struct {
```
Comment (lavalamp) [inline examples/apiserver/rest/reststorage.go]: Nit: Maybe "NotNamespacedScoped"?

**Comment by lavalamp** [comment]
LGTM, one name suggestion

> Sliced = a partial interface like:

Yeah, that's what I thought you meant, I just wasn't sure.

**Comment by caesarxuchao** [comment]
I've addressed the name suggestion. Applying the LGTM label per previous comment.

**Comment by bprashanth** [comment]
Apparently you need a rebase

**Comment by bprashanth** [comment]
@k8s-bot test this please https://github.com/kubernetes/kubernetes/issues/20306

**Comment by wojtek-t** [comment]
@caesarxuchao - thanks a lot for fixing it!

**Pull request merged**